WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/BlockProcessor.java

```diff
@@ -146,4 +120,33 @@ String processBlock(String block) {
      * @return A boolean value indicating whether or not the block contents should be replaced
      */
     abstract boolean processBlockContentDocument(Document document);
+
+    /**
+     * All concrete implementations must implement this method for the particular block type. The jsonAttributes object
+     * is a {@link JsonObject} parsed from the block header attributes. This object can be used to check for a match,
+     * and can be directly mutated if necessary.<br>
+     * <br>
+     * This method should return true to indicate success. Returning false will result in the block contents being
+     * unmodified.
+     *
+     * @param jsonAttributes the attributes object used to check for a match with the local id, and mutated if necessary
+     * @return
+     */
+    abstract boolean processBlockJsonAttributes(JsonObject jsonAttributes);
+
+    /**
+     * This method can be optionally overriden by concrete implementations to delegate further processing via recursion
+     * when {@link BlockProcessor#processBlockJsonAttributes(JsonObject)} returns false (i.e. the block did not match
+     * the local id being replaced). This is useful for implementing mutual recursion with
+     * {@link MediaUploadCompletionProcessor#processContent(String)} for block types that have media-containing blocks
+     * within their inner content.<br>
+     * <br>
+     * The default implementation provided is a NOOP that leaves the content of the block unchanged.
+     *
+     * @param block The raw block contents
+     * @return A string containing content with ids and urls replaced
+     */
+    String processInnerBlock(String block) {
+        return block;
```

Comment:
Is there a reason to not make the default processor process inner blocks? It will come surprisingly often that the blocks we care about are nested, for example, when people build with the full-site-editing layout blocks, or the Group block, or the Quote block, or the Gallery block…

Reply:
> Is there a reason to not make the default processor process inner blocks?

For the current implementation of the processors, they are not full-blown parsers, but part of their behavior resembles a parser. Instead, the processors are only concerned with a subset of block types (`image, video, etc.). To hopefully clarify: when `processContent` encounters a Group block, it does not match on that block, but if that block's inner contents contains a media-containing block, that will match and be processed further (via `processBlock`), while all content preceeding that match is treated as a raw string. This is a way to produce the desired output without the need to implement any serialization logic for all potential block types.